### PR TITLE
Implementation of Deque using array

### LINKED ISSUE #120: Array implementation of doubly ended queue (deque)
#### Description of the problem
I propose the implementation of a doubly ended queue (deque) using arrays. A deque allows insertions and deletions from both the front and the rear end.  The various methods available will be insert front, insert rear, delete front , delete rear in addition to returning the element at front or rear.

Hope this task will be assigned to me

### PR DESCRIPTION
**Pull request by Jahnavi-Jonnalagadda**
Added code for Implementation of Deque using array

#### References to other Issues or PRs or Relevant literature

Fixes #120 

#### Brief description of what is fixed or changed
Implemented Deque using arrays by adding popright() and appendleft() functions to the existing code.

#### Other comments
Code raises exception if the queue is not doubly ended

### CODE REVIEW

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 1]
```diff
@@ -41,22 +41,38 @@ class Queue(object):
...
-            return ArrayQueue(
-                kwargs.get('items', None),
-                kwargs.get('dtype', int))
+            if doubly_ended == True:
+                return ArrayQueue(
+                    kwargs.get('items', None),
+                    kwargs.get('dtype', int))
+            else:
+                raise TypeError("Queue is not doubly ended")
+
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
                return ArrayQueue(
                    kwargs.get('items', None),
                    kwargs.get('dtype', int),
                    kwargs.get('doubly_ended', False))
```

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 2]
```diff
@@ -41,22 +41,38 @@ class Queue(object):
     .. [1] https://en.wikipedia.org/wiki/Queue_(abstract_data_type)
     """
 
-    def __new__(cls, implementation='array', **kwargs):
+    def __new__(cls, implementation='array', doubly_ended = True, **kwargs):
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
    def __new__(cls, implementation='array', **kwargs):
        doubly_ended = kwargs.get('doubly_ended', False)
```

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 3]
```diff
@@ -41,22 +41,38 @@ class Queue(object):
...
-            return LinkedListQueue(
-                kwargs.get('items', None)
-            )
+            if doubly_ended == True:
+
+                return LinkedListQueue(
+                    kwargs.get('items', None)
+                )
+            else:
+                raise TypeError("Queue is not doubly ended")
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
            obj = LinkedListQueue(
                    kwargs.get('items', None)
                )
            obj.doubly_ended = doubly_ended
            return obj
```

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 4]
```diff
@@ -91,6 +107,16 @@ def append(self, x):
...
+    def appendleft(self, x):
+        if self.is_empty:
+            self.front = 0
+            self.rear = 0
+        elif self.front == 0:
+            self.front = self.items._size - 1
+        else:
+            self.front = self.front - 1
+        self.items.append(x)
+
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
    def appendleft(self, x):
        if not self.doubly_ended:
            raise TypeError("Queue is not doubly ended.")
        if self.is_empty:
            self.front = 0
            self.rear = 0
        elif self.front == 0:
            self.front = self.items._size - 1
        else:
            self.front = self.front - 1
        self.items.append(x)

```

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 5]
```diff
@@ -107,6 +133,19 @@ def popleft(self):
         self.items.delete(front_temp)
         return return_value
 
+    def popright(self):
+        if self.is_empty:
+            raise IndexError("Queue is empty")
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
    def popright(self):
        if not self.doubly_ended:
            raise TypeError("Queue is not doubly ended.")
        if self.is_empty:
            raise IndexError("Queue is empty")
```

**Comment by czgdp1807** [comment]
Nice work. I have made some suggestions to describe my point.

**Comment by czgdp1807** [comment]
Are you participating through GSSoC, 2020? Or would you like to count this contribution as RGSoC, 2020?

**Comment by Jahnavi-Jonnalagadda** [comment]
Thank You!

I have been part of GSSoC and will continue to contribute there.
But primary focus at this point is RGSoC as the application submission process is getting closed this month end.

Thank You for your support.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by czgdp1807** [comment]
Please add some tests to increase the diff coverage. 
P.S. If you are applying for RGSoC, 2020 in our project then please mention this PR in your application.

**Comment by Jahnavi-Jonnalagadda** [comment]
Yes, I will take this PR as reference and will reach out to you for any queries while filling the application form.

Btw, what is diff coverage? Sorry for the basic question.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by czgdp1807** [comment]
@Jahnavi-Jonnalagadda Are you working on it? There are some things which are to be addressed before this gets merged.

**Comment by czgdp1807** [comment]
> Btw, what is diff coverage? Sorry for the basic question.

The lines marked in red [here](https://codecov.io/gh/codezonediitj/pydatastructs/pull/174/diff) aren't executed even once by the tests i.e., those lines aren't tested so how can someone rely on them?

**Comment by Jahnavi-Jonnalagadda** [comment]
Hey, Sorry for this but could you please let me know what diff coverage is?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by czgdp1807** [comment]
Have you gone through https://github.com/codezonediitj/pydatastructs/pull/174#issuecomment-604040384 ?
If you still are not able to get it then please let me know.

**Comment by Jahnavi-Jonnalagadda** [comment]
Am I supposed to change the code of those lines in red?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

File: pydatastructs/miscellaneous_data_structures/queue.py [hunk 6]
```diff
@@ -91,6 +101,17 @@ def append(self, x):
...
+        if not self.doubly_ended:
+            raise TypeError("Queue is not doubly ended.")
+        if self.is_empty:
+            self.front = 0
+            self.rear = 0
+        elif self.front == 0:
+            self.front = self.items._size - 1
+        else:
+            self.front = self.front - 1
+        self.items.append(x)
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/queue.py]:
```suggestion
        self.items.append(x)

```

**Comment by czgdp1807** [comment]
> Am I supposed to change the code of those lines in red?

Tests are to be added for the lines in red in the following function, https://github.com/codezonediitj/pydatastructs/blob/6cc969c0c55ce348097025d917060d585f8054dd/pydatastructs/miscellaneous_data_structures/tests/test_queue.py#L17

In addition, `LinkedListQueue` should also have the definitions of `appendleft` and `popright` methods.

**Comment by Jahnavi-Jonnalagadda** [comment]
There are no lines in red in the link mentioned below?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by czgdp1807** [comment]
There red lines at https://codecov.io/gh/codezonediitj/pydatastructs/pull/174/diff
For analysing the coverage locally, see, https://github.com/codezonediitj/pydatastructs#testing

**Comment by Jahnavi-Jonnalagadda** [comment]
> > Am I supposed to change the code of those lines in red?
> 
> Tests are to be added for the lines in red in the following function,
> 
> https://github.com/codezonediitj/pydatastructs/blob/6cc969c0c55ce348097025d917060d585f8054dd/pydatastructs/miscellaneous_data_structures/tests/test_queue.py#L17
> 
> In addition, `LinkedListQueue` should also have the definitions of `appendleft` and `popright` methods.

**Comment by Jahnavi-Jonnalagadda** [comment]
Sorry for this, I've closed the PR by mistake. I actually thought of commenting.

**Comment by czgdp1807** [comment]
Any updates? After two days, `Please take over` label will be attached so that someone else can continue it.

**Comment by Jahnavi-Jonnalagadda** [comment]
I was referring to the tutorials you suggested. Will make the changes by tomorrow.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by HarsheetKakar** [comment]
Do you want `Queue` to be removed and implement `Deque` instead or you want a different class?

**Comment by czgdp1807** [comment]
As can be seen in the patch, an extra keyword argument is required to be added to enable `appendleft` and `popright` operations on the `Queue`. So, just small additions are needed to class `Queue` with some tests.

**Comment by HarsheetKakar** [comment]
> As can be seen in the patch, an extra keyword argument is required to be added to enable `appendleft` and `popright` operations on the `Queue`. So, just small additions are needed to class `Queue` with some tests.

so just the tests need to be updated?

**Comment by czgdp1807** [comment]
> so just the tests need to be updated?

Yes.

**Comment by HarsheetKakar** [comment]
> > so just the tests need to be updated?
> 
> Yes.

implementation is a bit off, need some time to correct it.

**Comment by HarsheetKakar** [comment]
if we use circular array implementation of `Deque` we cant use `DODA`, since it will add none at the end of the array and we want space between `rear` and `front`, we would have to first add circular dynamic array first and then move to `Deque`. Although [this](https://stackoverflow.com/a/6257048) comment suggests that `collections.deque` uses linked list instead of list, so if random access is not that important we can drop this idea.

**Comment by czgdp1807** [comment]
Well, then ignore the arrays in case of doubly ended queue, just raise an error if array implementation is asked for when doubly_ended=True.

**Comment by HarsheetKakar** [comment]
> Well, then ignore the arrays in case of doubly ended queue, just raise an error if array implementation is asked for when doubly_ended=True.

let me try working on circular dynamic queue. If it works then we might just have better implementation than python itself.

**Comment by czgdp1807** [comment]
Closing in favour of #247

**Pull request closed without merging**